x/archivedb/key_value_reader.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package archivedb
+
+import (
+	"bytes"
+
+	"github.com/ava-labs/avalanchego/database"
+)
+
+type dbHeightReader struct {
+	db                 *archiveDB
+	height             uint64
+	heightLastFoundKey uint64
+}
+
+// Builds a database.Iterator where the next value is the requested *value*. If
+// the value is not found a database.ErrNotFound is returned.
+//
+// This is a private function which helps Get() and Has() to share the key
+// lookup logic
+func (reader *dbHeightReader) getIteratorToValue(key []byte) (database.Iterator, error) {
+	internalKey := newInternalKey(key, reader.height)
+	iterator := reader.db.rawDB.NewIteratorWithStart(internalKey.Bytes())
+	keyLength := len(key)
+
+	reader.heightLastFoundKey = 0
+
+	for {
+		if !iterator.Next() {
+			// There is no available key with the requested prefix
+			return iterator, database.ErrNotFound
```

Comment:
I think we need to check `iterator.Error() == nil` before returning `database.ErrNotFound` here
Otherwise we may return `ErrNotFound` eg if the underlying db is closed during iteration.